File: android/vault-access/src/main/kotlin/org/secretary/browse/TrashModels.kt
package org.secretary.browse

/**
 * Read-only metadata for one trashed block. No secret material — the block name is plaintext in the
 * manifest; record content never leaves the core. Kotlin mirror of iOS `TrashedBlockInfo`.
 */
data class TrashedBlockInfo(
    val blockUuid: ByteArray,
    val blockName: String,
    val tombstonedAtMs: Long,
    val tombstonedBy: ByteArray,
) {
    /** Lowercase hex, no dashes — stable Compose list `key`. */
    val uuidHex: String get() = hexOfBytes(blockUuid)

    override fun equals(other: Any?): Boolean =
        other is TrashedBlockInfo &&
            blockUuid.contentEquals(other.blockUuid) &&
            blockName == other.blockName &&
            tombstonedAtMs == other.tombstonedAtMs &&
            tombstonedBy.contentEquals(other.tombstonedBy)

    override fun hashCode(): Int {
        var h = blockUuid.contentHashCode()
        h = 31 * h + blockName.hashCode()
        h = 31 * h + tombstonedAtMs.hashCode()
        h = 31 * h + tombstonedBy.contentHashCode()
        return h
    }
}

/** One trash entry eligible for retention auto-purge (preview only). Mirror of iOS `ExpiredEntryInfo`. */
data class ExpiredEntryInfo(
    val blockUuid: ByteArray,
    val tombstonedAtMs: Long,
    val ageMs: Long,
) {
    override fun equals(other: Any?): Boolean =
        other is ExpiredEntryInfo &&
            blockUuid.contentEquals(other.blockUuid) &&
            tombstonedAtMs == other.tombstonedAtMs &&
            ageMs == other.ageMs

    override fun hashCode(): Int {
        var h = blockUuid.contentHashCode()
        h = 31 * h + tombstonedAtMs.hashCode()
        h = 31 * h + ageMs.hashCode()
        return h
    }
}

/** Outcome of a single-block purge. Counts/classification only. Mirror of iOS `PurgeResultInfo`. */
data class PurgeResultInfo(
    val blockUuid: ByteArray,
    val wasShared: Boolean?,
    val recipientCount: Int?,
    val filesRemoved: Int,
) {
    override fun equals(other: Any?): Boolean =
        other is PurgeResultInfo &&
            blockUuid.contentEquals(other.blockUuid) &&
            wasShared == other.wasShared &&
            recipientCount == other.recipientCount &&
            filesRemoved == other.filesRemoved

    override fun hashCode(): Int {
        var h = blockUuid.contentHashCode()
        h = 31 * h + (wasShared?.hashCode() ?: 0)
        h = 31 * h + (recipientCount ?: 0)
        h = 31 * h + filesRemoved
        return h
    }
}

/** Aggregate outcome of an empty-trash batch. Counts only. Mirror of iOS `EmptyTrashReportInfo`. */
data class EmptyTrashReportInfo(
    val purgedCount: Int,
    val sharedCount: Int,
    val ownerOnlyCount: Int,
    val unknownCount: Int,
    val filesRemoved: Int,
    val filesFailed: Int,
)

/** Aggregate outcome of a retention auto-purge commit. Counts + echoed window. Mirror of iOS `RetentionReportInfo`. */
data class RetentionReportInfo(
    val purgedCount: Int,
    val sharedCount: Int,
    val ownerOnlyCount: Int,
    val unknownCount: Int,
    val filesRemoved: Int,
    val filesFailed: Int,
    val windowMs: Long,
)

/**
 * The vault-trash operations a Trash browser needs. Conformed by the `:kit` adapter
 * ([org.secretary.browse.UniffiVaultSession]) and by `FakeTrashPort` in tests. Kotlin mirror of the
 * iOS `TrashPort` protocol. Reports are returned (plumbed for #411) but the VM discards them.
 *
 * Reads ([listTrashedBlocks]/[expiredTrashEntries]/[defaultRetentionWindowMs]) are synchronous
 * in-memory manifest reads (no decryption). Writes are `suspend` — the real adapter offloads the
 * FFI write to the IO dispatcher, like [VaultSession.tombstoneRecord].
 */
interface TrashPort {
    /** All not-yet-purged trashed blocks, projected by name. */
    fun listTrashedBlocks(): List<TrashedBlockInfo>

    /** Retention preview for [windowMs] (adapter supplies `now`). Non-throwing (empty on wiped). */
    fun expiredTrashEntries(windowMs: Long): List<ExpiredEntryInfo>

    /** The frozen default retention window (90 days, in ms). */
    fun defaultRetentionWindowMs(): Long

    /** Restore the newest trashed copy of a block. */
    suspend fun restoreBlock(uuid: ByteArray)

    /** Permanently purge one trashed block. */
    suspend fun purgeBlock(uuid: ByteArray): PurgeResultInfo

    /** Permanently purge every currently-trashed block. */
    suspend fun emptyTrash(): EmptyTrashReportInfo

    /** Permanently purge every trashed block older than [windowMs]. */
    suspend fun autoPurgeExpired(windowMs: Long): RetentionReportInfo
}

File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccess/TrashModels.swift
import Foundation

/// Read-only metadata for one trashed block. No secret material — the block
/// name is plaintext in the manifest; record content never leaves the core.
public struct TrashedBlockInfo: Equatable {
    public let blockUuid: [UInt8]
    public let blockName: String
    public let tombstonedAtMs: UInt64
    public let tombstonedBy: [UInt8]

    public init(blockUuid: [UInt8], blockName: String,
                tombstonedAtMs: UInt64, tombstonedBy: [UInt8]) {
        self.blockUuid = blockUuid
        self.blockName = blockName
        self.tombstonedAtMs = tombstonedAtMs
        self.tombstonedBy = tombstonedBy
    }

    /// Lowercase hex, no dashes — stable SwiftUI list identity.
    public var uuidHex: String { blockUuid.map { String(format: "%02x", $0) }.joined() }
}

/// One trash entry eligible for retention auto-purge (preview only).
public struct ExpiredEntryInfo: Equatable {
    public let blockUuid: [UInt8]
    public let tombstonedAtMs: UInt64
    public let ageMs: UInt64

    public init(blockUuid: [UInt8], tombstonedAtMs: UInt64, ageMs: UInt64) {
        self.blockUuid = blockUuid
        self.tombstonedAtMs = tombstonedAtMs
        self.ageMs = ageMs
    }
}

/// Aggregate outcome of a retention auto-purge commit. Counts only.
public struct RetentionReportInfo: Equatable {
    public let purgedCount: UInt32
    public let sharedCount: UInt32
    public let ownerOnlyCount: UInt32
    public let unknownCount: UInt32
    public let filesRemoved: UInt32
    public let filesFailed: UInt32
    public let windowMs: UInt64

    public init(purgedCount: UInt32, sharedCount: UInt32, ownerOnlyCount: UInt32,
                unknownCount: UInt32, filesRemoved: UInt32, filesFailed: UInt32,
                windowMs: UInt64) {
        self.purgedCount = purgedCount
        self.sharedCount = sharedCount
        self.ownerOnlyCount = ownerOnlyCount
        self.unknownCount = unknownCount
        self.filesRemoved = filesRemoved
        self.filesFailed = filesFailed
        self.windowMs = windowMs
    }
}

/// Outcome of a single-block purge.
public struct PurgeResultInfo: Equatable {
    public let blockUuid: [UInt8]
    public let wasShared: Bool?
    public let recipientCount: UInt16?
    public let filesRemoved: UInt32

    public init(blockUuid: [UInt8], wasShared: Bool?,
                recipientCount: UInt16?, filesRemoved: UInt32) {
        self.blockUuid = blockUuid
        self.wasShared = wasShared
        self.recipientCount = recipientCount
        self.filesRemoved = filesRemoved
    }
}

/// Aggregate outcome of an empty-trash batch. Counts only.
public struct EmptyTrashReportInfo: Equatable {
    public let purgedCount: UInt32
    public let sharedCount: UInt32
    public let ownerOnlyCount: UInt32
    public let unknownCount: UInt32
    public let filesRemoved: UInt32
    public let filesFailed: UInt32

    public init(purgedCount: UInt32, sharedCount: UInt32, ownerOnlyCount: UInt32,
                unknownCount: UInt32, filesRemoved: UInt32, filesFailed: UInt32) {
        self.purgedCount = purgedCount
        self.sharedCount = sharedCount
        self.ownerOnlyCount = ownerOnlyCount
        self.unknownCount = unknownCount
        self.filesRemoved = filesRemoved
        self.filesFailed = filesFailed
    }
}

/// The vault-trash operations a Trash browser needs. Conformed by the
/// `SecretaryKit` adapter (`UniffiVaultSession`) and by `FakeTrashPort` in
/// tests. `AnyObject, Sendable` mirrors `VaultSession` (reference identity
/// for handle ownership; crosses the gate's async boundary).
public protocol TrashPort: AnyObject, Sendable {
    /// All not-yet-purged trashed blocks, projected by name.
    func listTrashedBlocks() throws -> [TrashedBlockInfo]
    /// Retention preview for `windowMs` (adapter supplies `now`). Non-throwing.
    func expiredTrashEntries(windowMs: UInt64) -> [ExpiredEntryInfo]
    /// The frozen default retention window (90 days).
    func defaultRetentionWindowMs() -> UInt64
    /// Restore the newest trashed copy of a block.
    func restoreBlock(uuid: [UInt8]) throws
    /// Permanently purge one trashed block.
    func purgeBlock(uuid: [UInt8]) throws -> PurgeResultInfo
    /// Permanently purge every currently-trashed block.
    func emptyTrash() throws -> EmptyTrashReportInfo
    /// Permanently purge every trashed block older than `windowMs`.
    func autoPurgeExpired(windowMs: UInt64) throws -> RetentionReportInfo
}
